### PR TITLE
YR/Add-mapping-and-different-authors/CIAC-9166

### PR DESCRIPTION
**Pull request by RosenbergYehuda**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-9316

## Description
In this PR:
- Modified the code to work with the 'xdr gitlab'.
- The flow now relies on commits instead of pipelines after a refactoring.
- The code has been improved.
- The mapping name functions has been added.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/scripts/common.py [hunk 1]
```diff
@@ -264,63 +265,65 @@ def person_in_charge(commit):
...
-        commits: list of gitlab commits
-        current_pipeline_sha: string
-        previous_pipeline_sha: string
+def are_pipelines_in_order(current_pipeline, previous_pipeline):
+    previous_pipeline_timestamp = parser.parse(previous_pipeline.created_at)
+    current_pipeline_timestamp = parser.parse(current_pipeline.created_at)
+    return current_pipeline_timestamp > previous_pipeline_timestamp
 
-    Returns:
-        boolean , the commit that triggered the current pipeline
```
Comment (JasBeilin) [inline Tests/scripts/common.py:284]: Why are we removing the docs? If its too much we can replace in a one line.

File: Tests/scripts/common.py [hunk 2]
```diff
@@ -264,63 +265,65 @@ def person_in_charge(commit):
...
-    if current_pipeline:
-        in_order, pivot_commit = are_pipelines_in_order_as_commits(commits, current_pipeline.sha, previous_pipeline.sha)
-        if in_order:
-            if previous_pipeline.status == 'success' and current_pipeline.status == 'failed':
-                return True, pivot_commit
-            if previous_pipeline.status == 'failed' and current_pipeline.status == 'success':
-                return False, pivot_commit
-    return None, None
+
+def is_pivot(current_pipeline, previous_pipeline):
```
Comment (JasBeilin) [inline Tests/scripts/common.py:279]: Please add docs here exlpaining why are we returning none for not ordered pipelines etc.

File: Tests/scripts/common.py [hunk 3]
```diff
@@ -264,63 +265,65 @@ def person_in_charge(commit):
...
+
+
+def get_reviewer(pr_url: str) -> str | None:
+    approved_reviewer = None
+    try:
+        # Extract the owner, repo, and pull request number from the URL
+        parts = pr_url.split("/")
+        repo_owner = parts[-4]
+        repo_name = parts[-3]
+        pr_number = parts[-1]
```
Comment (JasBeilin) [inline Tests/scripts/common.py:297]:
```suggestion
        repo_owner, repo_name, _, pr_number = pr_url.split("/")
```

File: Tests/scripts/gitlab_slack_notifier.py [hunk 4]
```diff
@@ -523,19 +529,31 @@ def main():
...
+                    if name == CONTENT_BOT:
+                        name = get_reviewer(pr)
+                    name = get_slack_user_name(name, options.name_mapping_path)
+                    msg = "broke" if pipeline_changed_status else "fixed"
+                    color = "danger" if pipeline_changed_status else "good"
+                    emoji = ":cry:" if pipeline_changed_status else ":muscle:"
+                    shame_message = (f"Hi @{name},  You {msg} the build! {emoji} ",
+                                     f" That was done in this {slack_link(pr,'PR.')}", color)
+
+                    computed_slack_channel = "test_slack_notifier_when_master_is_broken"
```
Comment (JasBeilin) [inline Tests/scripts/gitlab_slack_notifier.py:556]: Why not take this outside to a small helper func?

File: Tests/scripts/common.py [hunk 5]
```diff
@@ -221,34 +227,29 @@ def replace_escape_characters(sentence: str, replace_with: str = " ") -> str:
...
-    # Filter out pipelines that are not done
-    filtered_pipelines = [
-        pipeline for pipeline in pipelines if pipeline.status in ('success', 'failed')]
+                                       source='push', order_by='id', sort='asc')
 
-    return filtered_pipelines, commits
+    return pipelines, commits
 
 
 def person_in_charge(commit):
```
Comment (JasBeilin) [inline Tests/scripts/common.py:255]:
```suggestion
def get_person_in_charge(commit):
```

File: Tests/scripts/common.py [hunk 6]
```diff
@@ -264,63 +265,65 @@ def person_in_charge(commit):
         pr: The GitHub PR link for the Gitlab commit.
```
Comment (JasBeilin) [inline Tests/scripts/common.py:265]: The docs does not match the argument..

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 7]
```diff
@@ -1,147 +1,124 @@
...
+])
+def test_is_pivot(mocker, pipeline1_status, pipeline2_status, expected):
     """
-    Given a pipeline id, list of pipelines and commits
-    When the pipeline id is not in the list of pipelines
-    Then it should return None, None
+    Given:
+        - Two pipelines with their statuses.
+    When:
+        - is_pivot is called on them.
```
Comment (JasBeilin) [inline Tests/scripts/infrastructure_tests/common_test.py:62]:
```suggestion
        - Checking on status change.
```

**Review by JasBeilin — COMMENTED** [review]
Looks good! Now I am scared to approve since I might get blame :-P

File: Tests/scripts/infrastructure_tests/common_test.py (hunk 7 above)
Comment (RosenbergYehuda) [reply Tests/scripts/infrastructure_tests/common_test.py:62]: Done.

File: Tests/scripts/common.py (hunk 2 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:279]: fixed

File: Tests/scripts/common.py (hunk 3 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:297]: not as neet, but ok.
_1, _2,_3, repo_owner, repo_name,_4, pr_number = pr.split("/")

File: Tests/scripts/common.py (hunk 6 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:265]: fixed

File: Tests/scripts/common.py (hunk 5 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:255]: fixed

File: Tests/scripts/gitlab_slack_notifier.py (hunk 4 above)
Comment (RosenbergYehuda) [reply Tests/scripts/gitlab_slack_notifier.py:556]: Done.

File: Tests/scripts/common.py (hunk 1 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py]: My mistake.

File: Tests/scripts/common.py [hunk 8]
```diff
@@ -260,67 +265,104 @@ def person_in_charge(commit):
...
+        if previous_pipeline.status == 'failed' and current_pipeline.status == 'success':
+            return False
+    return None
+
+
+def get_reviewer(pr_url: str) -> str | None:
+    approved_reviewer = None
+    try:
+        # Extract the owner, repo, and pull request number from the URL
+        _1, _2,_3, repo_owner, repo_name,_4, pr_number = pr_url.split("/")
```
Comment (kobymeir) [inline Tests/scripts/common.py:317]: if some return variable isn't used, you can always write `_` for the return value.
```suggestion
        _ , _ ,_ , repo_owner, repo_name,_ , pr_number = pr_url.split("/")  
```

File: Tests/scripts/common.py [hunk 9]
```diff
@@ -260,67 +265,104 @@ def person_in_charge(commit):
...
+    except Exception as e:
+        logging.error(f"Failed to get reviewer for PR {pr_url}: {e}")
+    return approved_reviewer
+
+
+def get_slack_user_name(name: str, name_mapping_path: str) -> str:
+    with open(name_mapping_path) as map:
+        mapping = json.load(map)
+    # If the name is 'github-actions[bot]' (docker image update bot reviewer)  return the owner of that bot.
+        if name == 'github-actions[bot]':
```
Comment (kobymeir) [inline Tests/scripts/common.py:338]: move this `'github-actions[bot]'` to be a constant, as we did with the content-bot

File: Tests/scripts/gitlab_slack_notifier.py [hunk 10]
```diff
@@ -407,7 +400,14 @@ def construct_slack_msg(triggering_workflow: str,
 
     title += title_append
     title += f' - @{CI_SERVER_HOST}' if CI_SERVER_HOST else ''
-    return [{
+    slack_msg_start = []
+    if shame_message:
+        shame_title, shame_value, shame_color = shame_message
+        slack_msg_start.append({
+            "title": shame_title + "\n" + shame_value,
```
Comment (kobymeir) [inline Tests/scripts/gitlab_slack_notifier.py:407]: use f-string

File: Tests/scripts/infrastructure_tests/tests_data/test_mapping.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/tests_data/test_mapping.json:1]: pretty format the file

File: Tests/scripts/infrastructure_tests/tests_data/test_mapping.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+    "names":{
+    "Mike": "mike",
+    "Sam": "sam"},
+    "docker_images": {
+        "owner": "the owner"
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/tests_data/test_mapping.json:6]:
```suggestion
        "owner": "docker images file owner"
```
more explicit

File: Tests/scripts/infrastructure_tests/tests_data/test_mapping.json [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+    }
+}
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/tests_data/test_mapping.json:8]: missing new line at end of file

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 14]
```diff
@@ -1,147 +1,124 @@
...
+    requests_mock.get('https://api.github.com/repos/owner/repo/pulls/123/reviews', json=response)
+    result = get_reviewer(pr_url)
     assert result == expected
 
 
-def test_is_pivot_positive(mocker):
+@pytest.mark.parametrize('name, expected', [
+    ("Mike", "mike"),
+    ("Jon", "Jon"),
+    ("github-actions[bot]", "the owner")
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:109]: add `ids` to the parameterized.
@JasBeilin consider enforcing new parametrize rules.

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 15]
```diff
@@ -1,147 +1,124 @@
...
+    ('failed', 'success', False),
+    ('success', 'success', None),
+    ('failed', 'in progress', None)
+])
+def test_is_pivot(mocker, pipeline1_status, pipeline2_status, expected):
     """
-    Given a pipeline id, list of pipelines and commits
-    When the pipeline id is not in the list of pipelines
-    Then it should return None, None
+    Given:
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:59]: highly complex test, consider splitting to separate tests.
The scenario are not detailed enough.

**Review by kobymeir — CHANGES_REQUESTED** [review]

File: Tests/scripts/common.py (hunk 8 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:317]: fixed

File: Tests/scripts/common.py (hunk 9 above)
Comment (RosenbergYehuda) [reply Tests/scripts/common.py:338]: That make more sense. fixed.

File: Tests/scripts/gitlab_slack_notifier.py (hunk 10 above)
Comment (RosenbergYehuda) [reply Tests/scripts/gitlab_slack_notifier.py:407]: Done.

File: Tests/scripts/infrastructure_tests/common_test.py (hunk 14 above)
Comment (RosenbergYehuda) [reply Tests/scripts/infrastructure_tests/common_test.py:109]: Not furmiliar with that concept, and did not find examples online. Can you please explain what do you mean? in a comma like this?:
#1 ("Mike", "mike"),
#2   ("Jon", "Jon"),

File: Tests/scripts/infrastructure_tests/tests_data/test_mapping.json (hunk 12 above)
Comment (RosenbergYehuda) [reply Tests/scripts/infrastructure_tests/tests_data/test_mapping.json:6]: fixed

File: Tests/scripts/infrastructure_tests/tests_data/test_mapping.json (hunk 11 above)
Comment (RosenbergYehuda) [reply Tests/scripts/infrastructure_tests/tests_data/test_mapping.json:1]: Done

File: Tests/scripts/infrastructure_tests/common_test.py (hunk 14 above)
Comment (RosenbergYehuda) [reply Tests/scripts/infrastructure_tests/common_test.py]: This is new for me, cool!

File: Tests/scripts/common.py [hunk 16]
```diff
@@ -260,67 +264,106 @@ def person_in_charge(commit):
...
+                break
+    except Exception as e:
+        logging.error(f"Failed to get reviewer for PR {pr_url}: {e}")
+    return approved_reviewer
+
+
+def get_slack_user_name(name: str | None, name_mapping_path: str) -> str:
+    with open(name_mapping_path) as map:
+        mapping = json.load(map)
+    # If the name is the name of the 'docker image update bot' reviewer - return the owner of that bot.
```
Comment (kobymeir) [inline Tests/scripts/common.py:336]: comment - under indented.

File: Tests/scripts/gitlab_slack_notifier.py [hunk 17]
```diff
@@ -563,19 +556,22 @@ def main():
...
-            msg = "broke" if pipeline_changed_status else "fixed"
-            shame_message = (f"Hi @{name}, You {msg} the build.", f" That was done in this {slack_link(pr,'PR.')}")
-
-            computed_slack_channel = "test_slack_notifier_when_master_is_broken"
+        list_of_pipelines, list_of_commits = get_pipelines_and_commits(gitlab_client=gitlab_client,
+                                                                       project_id=project_id, look_back_hours=LOOK_BACK_HOURS)
+        current_commit = get_commit_by_sha(commit_sha, list_of_commits)
+        if current_commit:
+            current_commit_index = list_of_commits.index(current_commit)
+            previous_commit = list_of_commits[current_commit_index + 1]
```
Comment (kobymeir) [inline Tests/scripts/gitlab_slack_notifier.py:564]: that's strange that for the previous commit, you do a `+1` maybe worth mentioning that the order is descending.

File: Tests/scripts/gitlab_slack_notifier.py [hunk 18]
```diff
@@ -563,19 +556,22 @@ def main():
...
+            current_pipeline = get_pipeline_by_commit(current_commit, list_of_pipelines)
+            previous_pipeline = get_pipeline_by_commit(previous_commit, list_of_pipelines)
+            if current_pipeline and previous_pipeline:
+                pipeline_changed_status = is_pivot(current_pipeline, previous_pipeline)
+                logging.info(
+                    f"""Checking pipeline {pipeline_id},
+                    the commit is {current_commit} and the pipeline change status is: {pipeline_changed_status}""")
```
Comment (kobymeir) [inline Tests/scripts/gitlab_slack_notifier.py:571]: why use `"""` ? and not just `"` 

i don't think you meant it to have some many spaces and a new line in the string.

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 19]
```diff
@@ -1,147 +1,174 @@
...
+    result = get_person_in_charge(commit)
 
     assert result == expected
 
 
-def test_are_pipelines_in_order_as_commits_true(mocker):
+@pytest.mark.parametrize(('pipeline1_date, pipeline2_date, expected'), (
+    pytest.param('2020-01-01T00:00:00Z', '2020-01-02T00:00:00Z', False, id="pipelines not in order"),
+    pytest.param('2020-01-02T00:00:00Z', '2020-01-01T00:00:00Z', True, id="pipelines in order")))
+def test_are_pipelines_in_order(mocker, pipeline1_date, pipeline2_date, expected):
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:27]:
```suggestion
def test_pipelines_are_in_correct_order(mocker):
...
   old_pipeline = ...
   new_pipeline = ...
   assert are_pipelines_in_order(old_pipeline, new_pipeline) == True
   
 def test_pipelines_are_not_in_correct_order():
 ...
```
don't use parameterize

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 20]
```diff
@@ -1,147 +1,174 @@
...
-    Then it should return True and the commit that triggered the later pipeline
-
+    Given:
+        - Two pipelines with their created_at dates.
+    When:
+        - are_pipelines_in_order is called on them.
+    Then:
+        - It should return the expected result.
+          scenario 1: pipeline1.created_at > pipeline2.created_at -> False
+          scenario 2: pipeline1.created_at < pipeline2.created_at -> True
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:36]: please refer to https://www.businessanalysisexperts.com/gherkin-user-stories-given-when-then-examples/ 
for writing Given-When-Then

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 21]
```diff
@@ -1,147 +1,174 @@
...
+    result = are_pipelines_in_order(pipeline1, pipeline2)
 
     assert result == expected
 
 
-def test_are_pipelines_in_order_as_commits_false(mocker):
+@pytest.mark.parametrize(('current_pipeline_status, expected'), (
+    pytest.param('failed', True, id="negative pivot"),
+    pytest.param('success', None, id="no change")))
+def test_is_pivot__previously_pipeline_success(mocker, current_pipeline_status, expected):
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:54]: don't use parameterize, split into 2 tests

**Review by kobymeir — CHANGES_REQUESTED** [review]

File: Tests/scripts/common.py [hunk 22]
```diff
@@ -260,67 +264,106 @@ def person_in_charge(commit):
...
+
+def get_reviewer(pr_url: str) -> str | None:
+    approved_reviewer = None
+    try:
+        # Extract the owner, repo, and pull request number from the URL
+        _, _, _, repo_owner, repo_name, _, pr_number = pr_url.split("/")
+
+        # Get reviewers
+        reviews_url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/pulls/{pr_number}/reviews"
+        reviews_response = requests.get(reviews_url, verify=False)
```
Comment (kobymeir) [inline Tests/scripts/common.py:320]: `verify=False` - if you need it for local debugging, set it as a env var.

File: Tests/scripts/common.py (hunk 16 above)
Comment (kobymeir) [reply Tests/scripts/common.py:336]: issue still not resolved

File: Tests/scripts/common.py [hunk 23]
```diff
@@ -260,67 +264,106 @@ def person_in_charge(commit):
...
+    with open(name_mapping_path) as map:
+        mapping = json.load(map)
+    # If the name is the name of the 'docker image update bot' reviewer - return the owner of that bot.
+        if name == CONTENT_BOT_REVIEWER:
+            return mapping["docker_images"]["owner"]
+        else:
+            return mapping["names"].get(name, name)
+
+
+def get_commit_by_sha(commit_sha: str, list_of_commits: list) -> ProjectCommit | None:
```
Comment (kobymeir) [inline Tests/scripts/common.py:343]:
```suggestion
def get_commit_by_sha(commit_sha: str, list_of_commits: list[ProjectCommit]) -> ProjectCommit | None:
```

File: Tests/scripts/common.py [hunk 24]
```diff
@@ -260,67 +264,106 @@ def person_in_charge(commit):
...
+    return next((commit for commit in list_of_commits if commit.id == commit_sha), None)
+
+
+def get_pipeline_by_commit(commit, list_of_pipelines):
```
Comment (kobymeir) [inline Tests/scripts/common.py:347]: missing typing of args

File: Tests/scripts/common.py [hunk 25]
```diff
@@ -260,67 +264,106 @@ def person_in_charge(commit):
 
     Returns:
         name: The name of the commit author.
-        email: The email of the commit author.
         pr: The GitHub PR link for the Gitlab commit.
     """
     name = commit.author_name
-    email = commit.author_email
     pr_number = commit.title.split()[-1][2:-1]
```
Comment (kobymeir) [inline Tests/scripts/common.py:270]: can you explain the logic of `[2:-1]` why are you doing it?

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 26]
```diff
@@ -1,147 +1,262 @@
...
-def test_are_pipelines_in_order_as_commits_true(mocker):
+def test_pipelines_are_in_correct_order__false(mocker):
+    """
+    Given:
+        Two pipelines with different creation dates
+    When:
+        The function are_pipelines_in_order is called with the two pipelines
+    Then:
+        It should return False if the pipelines are not in order based on their creation dates
     """
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:32]:
```suggestion
    Given:
        Two pipelines that are out of order in respect to their creation time
    When:
        The function are_pipelines_in_order is called with the two pipelines
    Then:
        It should return False as the pipelines are out of order
    """
```
the Given-When-Then isn't good enough in the other UT as well.
hope my change in the text explains what i'm doing differently, if not, approach me.

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 27]
```diff
@@ -1,147 +1,262 @@
...
-    pipelines = [
-        mocker.Mock(id=1, status='success'),
-        mocker.Mock(id=2, status='failed')
-    ]
-    commit = mocker.Mock()
-    commits = [commit]
+    pr_url = 'https://github.com/owner/repo/pull/123'
+    response = [{"Jon": "test", "state": "test", "user": {"login": "Jon"}},
+                {"Jane Doe": "test", "state": "APPROVED", "user": {"login": "Jane Doe"}}]
+    expected = "Jane Doe"
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:196]: no need for the variable assignment
```suggestion
```

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 28]
```diff
@@ -1,147 +1,262 @@
...
+    requests_mock.get('https://api.github.com/repos/owner/repo/pulls/123/reviews', json=response)
+    result = get_reviewer(pr_url)
+    assert result == expected
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:199]:
```suggestion
    assert result == "Jane Doe"
```

File: Tests/scripts/infrastructure_tests/common_test.py [hunk 29]
```diff
@@ -1,147 +1,262 @@
...
+        - A PR URL with two reviewers who approved.
+    When:
+        - get_reviewer is called on it.
+    Then:
+        - It should return the first reviewer's name.
+    """
+    pr_url = 'https://github.com/owner/repo/pull/123'
+    response = [{"Jon": "test", "state": "APPROVED", "user": {"login": "Jon"}},
+                {"Jane Doe": "test", "state": "APPROVED", "user": {"login": "Jane Doe"}}]
+    expected = "Jon"
```
Comment (kobymeir) [inline Tests/scripts/infrastructure_tests/common_test.py:214]: same here

**Review by kobymeir — CHANGES_REQUESTED** [review]

**Review by kobymeir — APPROVED** [review]

**Pull request merged**